src/tox_travis.py

```diff
@@ -99,28 +119,121 @@ def get_default_envlist(version):
     return guess_python_env()
 
 
-def get_desired_envs(config, version):
-    """Get the expanded list of desired envs."""
-    travis_section = config.sections.get('tox:travis', {})
-    default_envlist = get_default_envlist(version)
-    return split_env(travis_section.get(version, default_envlist))
+def get_desired_factors(config):
+    """Get the list of desired envs per declared factor.
+
+    Look at all the accepted configuration locations, and give a list
+    of envlists, one for each Travis factor found.
+
+    Look in the ``[travis]`` section for the known Travis factors,
+    which are backed by environment variable checking behind the
+    scenes, but provide a cleaner interface.
+
+    Also look for the ``[tox:travis]`` section, which is deprecated,
+    and treat it as an additional ``python`` key from the ``[travis]``
+    section.
+
+    Finally, look for factors based directly on environment variables,
+    listed in the ``[travis:env]`` section. Configuration found in the
+    ``[travis]`` and ``[tox:travis]`` sections are converted to this
+    form under the hood, and are considered in the same way.
+
+    Special consideration is given to the ``python`` factor. If this
+    factor is set in the environment, then an appropriate configuration
+    will be provided automatically if no manual configuration is
+    provided.
+
+    To allow for the most flexible processing, the envlists provided
+    by each factor are not combined after they are selected, but
+    instead returned as a list of envlists, and expected to be
+    combined as and when appropriate by the caller. This allows for
+    special handling based on the number of factors that were found
+    to apply to this environment.
+    """
+    # Find configuration based on known travis factors
+    travis_section = config.sections.get('travis', {})
+    found_factors = [
+        (factor, parse_dict(travis_section[factor]))
+        for factor in TRAVIS_FACTORS
+        if factor in travis_section
+    ]
+
+    # Backward compatibility with the old tox:travis section
+    if 'tox:travis' in config.sections:
+        print('The [tox:travis] section is deprecated in favor of'
+              ' the "python" key of the [travis] section.', file=sys.stderr)
+        found_factors.append(('python', config.sections['tox:travis']))
```

Comment:
Should this check to see if `python` is already in `found_factors`? It should be nonsensical (probably?), but it would imply that the user had both a `tox:travis` section and a `travis` section w/ `python` factor.